app/webdav/webdav.go

```diff
@@ -0,0 +1,407 @@
+	if err != nil {
+		return fmt.Errorf("unable to describe blob ref %s: %w", n.br, err)
+	}
+	db := des.Meta.Get(n.br)
+
+	switch db.CamliType {
+	case schema.TypePermanode:
+		// TODO
+		return fmt.Errorf("unable to refresh permanodes %s: %w", n.br, err)
+	case schema.TypeFile:
+		n.static = true
+		n.fi = webdavFsFileInfo{
+			isDir:   false,
+			name:    db.File.FileName,
+			size:    db.File.Size,
+			mode:    0400,
+			modTime: modtimeFromFileInfo(db.File),
+		}
+	case schema.TypeDirectory:
+		n.static = true
+		n.fi = webdavFsFileInfo{
+			isDir:   true,
+			name:    db.Dir.FileName,
+			size:    db.Dir.Size,
+			mode:    0400,
+			modTime: modtimeFromFileInfo(db.Dir),
+		}
+		n.sub = make(map[string]*webdavFsNode)
+		for _, m := range db.DirMembers() {
+			dmc := des.Meta.Get(m.BlobRef)
+			var fi *camtypes.FileInfo
+			if dmc.File != nil {
+				fi = dmc.File
+			} else if dmc.Dir != nil {
+				fi = dmc.Dir
+			}
+			if fi == nil {
+				continue
+			}
+			n.sub[fi.FileName] = &webdavFsNode{br: m.BlobRef}
+		}
+	}
+	n.lastRefreshed = time.Now()
+	return nil
+}
+
+type webdavFsFile struct {
+	n   *webdavFsNode
+	pos int
+
+	// file
+	r *schema.FileReader
+
+	// dir
+	dentries []os.FileInfo
+}
+
+var (
+	_ webdav.File = (*webdavFsFile)(nil)
+)
+
+func (f *webdavFsFile) isDir() bool {
+	return f.r == nil
+}
+
+func (f *webdavFsFile) Seek(offset int64, whence int) (int64, error) {
+	npos := f.pos
+	switch whence {
+	case io.SeekStart:
+		npos = int(offset)
+	case io.SeekCurrent:
+		npos += int(offset)
+	case io.SeekEnd:
+		npos = int(f.n.fi.Size()) + int(offset)
+	default:
+		npos = -1
+	}
+	if npos < 0 {
+		return 0, os.ErrInvalid
+	}
+	f.pos = npos
+	return int64(f.pos), nil
+}
+
+func (f *webdavFsFile) ETag(ctx context.Context) (string, error) {
+	return f.n.br.Digest(), nil
+}
+
+func (f *webdavFsFile) Read(p []byte) (int, error) {
+	if f.isDir() {
+		return 0, os.ErrInvalid
+	}
+	n, err := f.r.ReadAt(p, int64(f.pos))
+	f.pos += n
+	return n, err
+}
+
+func (f *webdavFsFile) Write(p []byte) (int, error) {
+	return 0, os.ErrPermission
+}
+
+func (f *webdavFsFile) Close() error {
+	if f.isDir() {
+		return nil
+	}
+	return f.r.Close()
+}
+
+func (f *webdavFsFile) Readdir(count int) ([]os.FileInfo, error) {
+	if !f.isDir() {
+		return nil, os.ErrInvalid
+	}
+	old := f.pos
+	if old >= len(f.dentries) {
+		if count > 0 {
+			return nil, io.EOF
+		}
+		return nil, nil
+	}
+	if count > 0 {
+		f.pos += count
+		if f.pos > len(f.dentries) {
+			f.pos = len(f.dentries)
+		}
+	} else {
+		f.pos = len(f.dentries)
+		old = 0
+	}
+	return f.dentries[old:f.pos], nil
+}
+
+func modtimeFromFileInfo(fi *camtypes.FileInfo) time.Time {
+	t := time.Now()
+	if fi.Time != nil {
+		t = fi.Time.Time()
+	}
+	if fi.ModTime != nil {
+		t = fi.ModTime.Time()
+	}
+	return t
+}
+
+func (f *webdavFsFile) Stat() (os.FileInfo, error) {
+	return f.n.fi, nil
+}
+
+type webdavFsFileInfo struct {
```

Comment:
all these "webdav" type prefixes are redundant with the package name. Could just call this "fsFileInfo" ?

Reply:
:heavy_check_mark: